zerver/views/messages.py

```diff
@@ -172,11 +172,15 @@ def add_term(self, query: Query, term: Dict[str, Any]) -> Query:
         return method(query, operand, maybe_negate)
 
     def by_has(self, query: Query, operand: str, maybe_negate: ConditionTransform) -> Query:
-        if operand not in ['attachment', 'image', 'link']:
-            raise BadNarrowOperator("unknown 'has' operand " + operand)
-        col_name = 'has_' + operand
-        cond = column(col_name)
-        return query.where(maybe_negate(cond))
+        if operand in ['attachment', 'image', 'link']:
+            col_name = 'has_' + operand
+            cond = column(col_name)
+            return query.where(maybe_negate(cond))
+        elif operand == 'reaction':
+            message_id = column('message_id')
+            Reaction = table('zerver_reaction')
+            return query.where(message_id.in_(select([message_id]).select_from(Reaction)))
+        raise BadNarrowOperator("unknown 'has' operand " + operand)
```

Comment:
Can you check what postgres query this generates?  I think the easiest way to get that is to write a backend test (which we want to do anyway) and use `with queries_captured() as queries:` to capture them.

And then ideally we'll do an `EXPLAIN ANALYZE` on that query to make sure it has a sensible query plan.